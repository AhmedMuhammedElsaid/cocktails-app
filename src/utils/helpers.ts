/*
 @params object
 return object with no null ,undefined , empty string "cleaned object"
*/
export function cleanObject(object: { [x: string]: any }) {
  if (!object) return {};
  if (typeof object === "object") {
    for (let propName in object) {
      if (
        object[propName] === null ||
        object[propName] === undefined ||
        object[propName]?.length === 0
      ) {
        delete object[propName];
      }
    }
  }
  return object;
}
