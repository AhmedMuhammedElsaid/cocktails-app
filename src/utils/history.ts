import { createBrowserHistory } from "history";

const history = createBrowserHistory();
// method to redirect the user to error page in case any error happens while fetching the data from the server
export const redirectToErrorPage = () => {
  history.push(`/pagenotfound`);
  history.go(0);
};
export default history;
