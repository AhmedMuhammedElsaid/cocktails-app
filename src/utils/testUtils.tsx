import { render, RenderResult } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";

// this is a handy function that I would utilize for any component
// that relies on the router being in context
export const customRender = (component: JSX.Element): RenderResult => {
  return {
    ...render(<Router>{component}</Router>),
  };
};
