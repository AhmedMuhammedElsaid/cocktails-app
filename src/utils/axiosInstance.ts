import axios from "axios";
import { redirectToErrorPage } from './history';
import { toast } from "react-toastify";

const axiosInstance = axios.create({
  // setting the default uri for this instance
  baseURL: process.env.REACT_APP_BASE_URI,
});

// setting language  configuration on every single request we send to the BE
axiosInstance.interceptors.request.use(
  (request: any) => {
    return {
      ...request,
      headers: {
        ...request.headers,
        // setting the header lang german "de"
        "Accept-Language": "de",
      },
    };
  },
  (error) => {
    return Promise.reject(error);
  }
);

axiosInstance.interceptors.response.use(
  (response) => {
    return response.data;
  },
  (error) => {
    switch (error?.response?.status) {
      case 400:
        return {
          data: null,
          hasError: true,
          error: [error.response.data],
        };
      case 401:
        toast.error("Unauthorized Call Access");
        break;
      case 404:
        toast.error("404 Not Found");
        break;
      case 500:
        toast.error("Server Is Down");
        break;
    }
    redirectToErrorPage()
    return Promise.reject(error);
  }
);
export default axiosInstance;
