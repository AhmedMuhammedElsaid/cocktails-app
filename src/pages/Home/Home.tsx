import React, {
  useState,
  useCallback,
  useEffect,
  FunctionComponent,
} from "react";
import { Divider, Grid } from "@material-ui/core";
import { Fade } from "react-reveal";
import debounce from "lodash.debounce";
import CocktailInput from "components/CocktailInput";
import CocktailList from "../../components/CocktailsList";
import { AxiosResponse, cocktailSchema } from "utils/types";
import CocktailCard from "components/CocktailCard";
import Loader from "components/Loader";

import axiosInstance from "utils/axiosInstance";
import { toast } from "react-toastify";

const Home: FunctionComponent = () => {
  const [cocktailName, setCocktailName] = useState<string>("");
  const [isLoading, setIsLoading] = useState(false);
  const [randomCocktail, setRandomCocktail] = useState<null | cocktailSchema>(
    null
  );
  const [currentCocktails, setCurrentCocktails] = useState<
    cocktailSchema[] | null
  >(null);

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>): void => {
    debouncedSearch(event.target.value);
    setCocktailName(event.target.value);
  };
  const getRandomCocktail = async () => {
    const result: AxiosResponse = await axiosInstance.get("/random.php");

    const { drinks } = result;
    if (drinks) setRandomCocktail(drinks?.[0]);
  };
  const handleSearch = async (cocktailInitialLetter: string) => {
    if (cocktailInitialLetter.length === 1) {
      setIsLoading(true);
      const result: AxiosResponse = await axiosInstance.get(
        `/search.php?f=${cocktailInitialLetter}`
      );
      const { drinks } = result;
      if (drinks) {
        setCurrentCocktails(drinks);
      }
      setIsLoading(false);
    } else {
      toast.error("Please Enter Only One Letter ");
      setCurrentCocktails(null);
    }
  };
  const debouncedSearch = useCallback(
    debounce((value: string) => handleSearch(value), 500),
    []
  );
  useEffect(() => {
    getRandomCocktail();
  }, []);

  return (
    <Fade top duration={2000} distance="40px">
      <Grid container>
        <Grid item sm={12}>
          <CocktailInput value={cocktailName} handleChange={handleChange} />
        </Grid>
        <Grid item sm={12}>
          {isLoading ? (
            <Loader />
          ) : (
            <Fade top duration={2000} distance="40px">
              <CocktailList cocktails={currentCocktails} />
            </Fade>
          )}
        </Grid>
        <Grid container alignItems="center" justifyContent="center">
          <Grid item sm={3} className="Random__Cocktails">
            <Divider variant="fullWidth" />
            {randomCocktail ? (
              <Fade bottom duration={2000} distance="40px">
                <h2>Der zuf√§llige Cocktail</h2>
                <CocktailCard cocktail={randomCocktail} />
              </Fade>
            ) : (
              <Loader />
            )}
          </Grid>
        </Grid>
      </Grid>
    </Fade>
  );
};

export default Home;
