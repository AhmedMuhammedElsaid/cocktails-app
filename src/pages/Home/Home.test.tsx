import Home from "./Home";
import { customRender } from "utils/testUtils";
import "@testing-library/jest-dom";
import { cleanup, fireEvent } from "@testing-library/react";
const cocktailInput = "cocktail-input-testid",
  loadingSpinner = "loader-spinner-testid";
describe("test default render to the home page", () => {
  afterEach(cleanup);
  it("should render home page without crashing", () => {
    const { asFragment, getByTestId } = customRender(<Home />);
    expect(getByTestId(cocktailInput)).toBeInTheDocument();
    expect(getByTestId(loadingSpinner)).toBeInTheDocument();
    fireEvent.change(getByTestId(cocktailInput), { target: { value: "f" } });
    expect(getByTestId(cocktailInput)).toHaveValue("f");
    expect(asFragment()).toMatchSnapshot();
  });
});
