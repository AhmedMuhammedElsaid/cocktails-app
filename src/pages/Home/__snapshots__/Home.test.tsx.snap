// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test default render to the home page should render home page without crashing 1`] = `
<DocumentFragment>
  <div
    class="react-reveal"
  >
    <div
      class="MuiGrid-root MuiGrid-container"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-12"
      >
        <div
          class="Container"
        >
          <div
            class="Container__Head"
          >
            <div>
              <h2>
                Cocktail Searcher
              </h2>
              <p>
                browse through all cocktails based on initial letter 
                <br />
                cocktails with the initial letter will show up immediately 
              </p>
            </div>
          </div>
          <div
            class="Container__Input__Icon"
          >
            <input
              autocomplete="false"
              data-testid="cocktail-input-testid"
              name="name"
              placeholder="search by initial cocktail letter here...."
              type="search"
              value="f"
            />
            <img
              alt="search icon"
              src="searchIcon.png"
            />
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-12"
      >
        <div
          class="react-reveal"
        >
          <div
            class="MuiGrid-root Container"
            data-testid="card-holder-testid"
          />
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-content-xs-center"
      >
        <div
          class="MuiGrid-root Random__Cocktails MuiGrid-item MuiGrid-grid-sm-3"
        >
          <hr
            class="MuiDivider-root"
          />
          <div
            class="Loader"
            data-testid="loader-spinner-testid"
          >
            <div
              class="MuiCircularProgress-root MuiCircularProgress-indeterminate"
              role="progressbar"
              style="width: 40px; height: 40px;"
            >
              <svg
                class="MuiCircularProgress-svg"
                viewBox="22 22 44 44"
              >
                <circle
                  class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
                  cx="44"
                  cy="44"
                  fill="none"
                  r="20.2"
                  stroke-width="3.6"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
